File: packages/shared/src/validation/endpoints-validation.ts
import type { EndpointsConfig, Endpoint, EndpointType } from "@shared/types";

export const ENDPOINT_TYPES: { value: EndpointType; label: string }[] = [
  { value: "prometheus", label: "Prometheus" },
  { value: "loki", label: "Loki" },
  { value: "kubernetes", label: "Kubernetes" },
  { value: "aws", label: "AWS (Cloud)" },
];

export type EditableEndpoint = {
  id: string;
  name: string;
  type: EndpointType;
  // prometheus / loki
  url: string;
  http_auth_mode: string; // none | basic | bearer
  username: string;
  password: string;
  bearer_token: string;
  // kubernetes
  kube_context: string;
  api_server: string;
  kube_token: string;
  ca_cert: string;
  // aws
  region: string;
  aws_auth_mode: string; // default | assume_role | keys
  role_arn: string;
  access_key_id: string;
  secret_access_key: string;
};

export function blankEndpoint(type: EndpointType, id: string): EditableEndpoint {
  return {
    id,
    name: "",
    type,
    url: "",
    http_auth_mode: "none",
    username: "",
    password: "",
    bearer_token: "",
    kube_context: "",
    api_server: "",
    kube_token: "",
    ca_cert: "",
    region: "",
    aws_auth_mode: "default",
    role_arn: "",
    access_key_id: "",
    secret_access_key: "",
  };
}

function isHttpUrl(u: string): boolean {
  const t = u.trim();
  if (!t) return false;
  try {
    const p = new URL(t);
    return p.protocol === "http:" || p.protocol === "https:";
  } catch {
    return false;
  }
}

/** Build the API payload; masked secrets (***) are sent verbatim and the backend
 * restores them from the stored registry. */
export function toEndpointsConfig(eps: EditableEndpoint[]): EndpointsConfig {
  return {
    endpoints: eps.map((e) => {
      const out: Endpoint = { name: e.name.trim(), type: e.type };
      if (e.type === "prometheus" || e.type === "loki") {
        out.url = e.url.trim();
        const mode = e.http_auth_mode || "none";
        out.auth = { mode };
        if (mode === "basic") {
          out.auth.username = e.username;
          out.auth.password = e.password;
        } else if (mode === "bearer") {
          out.auth.token = e.bearer_token;
        }
      } else if (e.type === "kubernetes") {
        if (e.kube_context.trim()) out.kube_context = e.kube_context.trim();
        if (e.api_server.trim()) out.api_server = e.api_server.trim();
        if (e.kube_token) out.token = e.kube_token;
        if (e.ca_cert.trim()) out.ca_cert = e.ca_cert;
      } else if (e.type === "aws") {
        if (e.region.trim()) out.region = e.region.trim();
        const mode = e.aws_auth_mode || "default";
        out.auth = { mode };
        if (mode === "assume_role") {
          out.auth.role_arn = e.role_arn;
        } else if (mode === "keys") {
          out.auth.access_key_id = e.access_key_id;
          out.auth.secret_access_key = e.secret_access_key;
        }
      }
      return out;
    }),
  };
}

export function fromEndpointsConfig(cfg: EndpointsConfig | undefined): EditableEndpoint[] {
  return (cfg?.endpoints ?? []).map((e, i) => ({
    id: `ep-${i}-${e.name ?? ""}`,
    name: e.name ?? "",
    type: (e.type ?? "prometheus") as EndpointType,
    url: e.url ?? "",
    http_auth_mode: e.auth?.mode ?? "none",
    username: e.auth?.username ?? "",
    password: e.auth?.password ?? "",
    bearer_token: e.auth?.token ?? "",
    kube_context: e.kube_context ?? "",
    api_server: e.api_server ?? "",
    kube_token: e.token ?? "",
    ca_cert: e.ca_cert ?? "",
    region: e.region ?? "",
    aws_auth_mode: e.auth?.mode ?? "default",
    role_arn: e.auth?.role_arn ?? "",
    access_key_id: e.auth?.access_key_id ?? "",
    secret_access_key: e.auth?.secret_access_key ?? "",
  }));
}

export type EndpointValidation = { valid: boolean; eps: Record<string, string> };

export function validateEndpoints(eps: EditableEndpoint[]): EndpointValidation {
  const result: EndpointValidation = { valid: true, eps: {} };
  const seen = new Set<string>();

  for (const e of eps) {
    let err = "";
    const name = e.name.trim();
    if (!name) err = "Name is required.";
    else if (seen.has(name)) err = `Duplicate name "${name}".`;
    else seen.add(name);

    if (e.type === "prometheus" || e.type === "loki") {
      if (!isHttpUrl(e.url)) err = err || "URL must be an http(s) URL.";
      if (e.http_auth_mode === "basic" && !e.username.trim()) err = err || "Basic auth requires a username.";
      if (e.http_auth_mode === "bearer" && !e.bearer_token.trim()) err = err || "Bearer auth requires a token.";
    } else if (e.type === "kubernetes") {
      if (!e.kube_context.trim() && !e.api_server.trim()) {
        err = err || "Set a kube-context, or api-server + token (or leave empty for in-cluster).";
      }
      if (e.api_server.trim()) {
        if (!isHttpUrl(e.api_server)) err = err || "API server must be an http(s) URL.";
        if (!e.kube_token.trim()) err = err || "API server requires a token.";
      }
    } else if (e.type === "aws") {
      if (e.aws_auth_mode === "assume_role" && !e.role_arn.trim()) err = err || "Assume-role requires a role ARN.";
      if (e.aws_auth_mode === "keys" && !(e.access_key_id.trim() && e.secret_access_key.trim())) {
        err = err || "Keys mode requires an access key ID and secret.";
      }
    }

    if (err) {
      result.valid = false;
      result.eps[e.id] = err;
    }
  }

  return result;
}

/** name → type map, used by the environments page to populate + validate dropdowns. */
export function endpointsByType(eps: EditableEndpoint[]): Record<string, string> {
  const map: Record<string, string> = {};
  for (const e of eps) if (e.name.trim()) map[e.name.trim()] = e.type;
  return map;
}

File: alert-agent/services/endpoints_validation.py
"""Validate the named endpoint registry submitted via the Web UI.

Secrets are merged in (``***`` → stored value) before this runs, so validation
sees real values. See ``api/config_api.merge_endpoint_secrets``.
"""

from typing import Any
from urllib.parse import urlparse

ENDPOINT_TYPES = ("prometheus", "loki", "kubernetes", "aws")
_HTTP_AUTH_MODES = ("none", "basic", "bearer")
_AWS_AUTH_MODES = ("default", "assume_role", "keys")


def _is_http_url(url: str) -> bool:
    url = (url or "").strip()
    if not url:
        return False
    try:
        parsed = urlparse(url)
    except Exception:
        return False
    return parsed.scheme in ("http", "https") and bool(parsed.netloc)


def _validate_http(prefix: str, ep: dict, errors: list[str]) -> None:
    if not _is_http_url(str(ep.get("url") or "")):
        errors.append(f"{prefix}: url must be an http(s) URL.")
    auth = ep.get("auth") or {}
    if not isinstance(auth, dict):
        errors.append(f"{prefix}: auth must be an object.")
        return
    mode = str(auth.get("mode") or "none").strip().lower()
    if mode not in _HTTP_AUTH_MODES:
        errors.append(f"{prefix}: auth.mode must be one of {', '.join(_HTTP_AUTH_MODES)}.")
    elif mode == "basic" and not str(auth.get("username") or "").strip():
        errors.append(f"{prefix}: basic auth requires a username.")
    elif mode == "bearer" and not str(auth.get("token") or "").strip():
        errors.append(f"{prefix}: bearer auth requires a token.")


def _validate_kubernetes(prefix: str, ep: dict, errors: list[str]) -> None:
    ctx = str(ep.get("kube_context") or "").strip()
    api_server = str(ep.get("api_server") or "").strip()
    token = str(ep.get("token") or "").strip()
    if not ctx and not api_server:
        errors.append(f"{prefix}: set kube_context, or api_server + token (or leave all empty for in-cluster).")
        return
    if api_server:
        if not _is_http_url(api_server):
            errors.append(f"{prefix}: api_server must be an http(s) URL.")
        if not token:
            errors.append(f"{prefix}: api_server requires a token.")


def _validate_aws(prefix: str, ep: dict, errors: list[str]) -> None:
    auth = ep.get("auth") or {}
    if not isinstance(auth, dict):
        errors.append(f"{prefix}: auth must be an object.")
        return
    mode = str(auth.get("mode") or "default").strip().lower()
    if mode not in _AWS_AUTH_MODES:
        errors.append(f"{prefix}: auth.mode must be one of {', '.join(_AWS_AUTH_MODES)}.")
    elif mode == "assume_role" and not str(auth.get("role_arn") or "").strip():
        errors.append(f"{prefix}: assume_role requires a role_arn.")
    elif mode == "keys" and not (
        str(auth.get("access_key_id") or "").strip() and str(auth.get("secret_access_key") or "").strip()
    ):
        errors.append(f"{prefix}: keys mode requires access_key_id and secret_access_key.")


def validate_endpoints_body(body: dict[str, Any]) -> list[str]:
    """Return a list of human-readable validation errors (empty if valid)."""
    errors: list[str] = []

    endpoints = body.get("endpoints")
    if endpoints is None or not isinstance(endpoints, list):
        errors.append("endpoints must be a list.")
        return errors

    seen: set[str] = set()
    for i, ep in enumerate(endpoints, start=1):
        if not isinstance(ep, dict):
            errors.append(f"Endpoint {i} must be an object.")
            continue

        name = str(ep.get("name") or "").strip()
        etype = str(ep.get("type") or "").strip().lower()
        prefix = f"Endpoint {i}" + (f" ({name})" if name else "")

        if not name:
            errors.append(f"{prefix}: name is required.")
        elif name in seen:
            errors.append(f"{prefix}: duplicate name {name!r}.")
        else:
            seen.add(name)

        if etype not in ENDPOINT_TYPES:
            errors.append(f"{prefix}: type must be one of {', '.join(ENDPOINT_TYPES)}.")
            continue

        if etype in ("prometheus", "loki"):
            _validate_http(prefix, ep, errors)
        elif etype == "kubernetes":
            _validate_kubernetes(prefix, ep, errors)
        elif etype == "aws":
            _validate_aws(prefix, ep, errors)

    return errors
